src/coreclr/vm/exceptionhandling.cpp

```diff
@@ -5857,7 +5857,7 @@ struct _Unwind_Exception;
 // This is a personality routine for TheUMEntryPrestub and UMThunkStub Unix asm stubs.
 // An exception propagating through these stubs is an unhandled exception.
 // This function dumps managed stack trace and terminates the current process.
-EXTERN_C _Unwind_Reason_Code
+DLLEXPORT EXTERN_C _Unwind_Reason_Code
```

Comment:
```suggestion
EXTERN_C _Unwind_Reason_Code
UnhandledExceptionHandlerUnix(
                IN int version,
                IN _Unwind_Action action,
                IN uint64_t exceptionClass,
                IN struct _Unwind_Exception *exception,
                IN struct _Unwind_Context *context
              );
              
_Unwind_Reason_Code
```
DLLEXPORT does not look appropriate here. This symbol is not meant to be used outside libcoreclr.so.

Can the problem be fixed by declaring the method separately?

Reply:
lld-15 was giving an error on riscv64 target: `ld.lld: error: cannot preempt symbol: UnhandledExceptionHandlerUnix`, this patch fixed the error and build succeeded.

Reply:
Is this trying to work around a bug in the compiler?

Reply:
Based on my understanding of this article https://maskray.me/blog/2021-01-09-copy-relocations-canonical-plt-entries-and-protected, it is due to hardenining in linker around `visibility`. It only works with `__attribute__((visibility("default")))` and fails with `hidden` and `protected`.

```sh
# this PR branch without DLLEXPORT change
$ docker run --rm -it -v$(pwd):/runtime -e ROOTFS_DIR=/crossrootfs/riscv64 \
     mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-22.04-cross-riscv64-20230420175531-3fc5553 \
     /runtime/build.sh clr -cross -a riscv64

...
  [100%] Linking CXX executable singlefilehost
  ld.lld: error: cannot preempt symbol: UnhandledExceptionHandlerUnix
  >>> defined in ../vm/wks/CMakeFiles/cee_wks_core.dir/__/exceptionhandling.cpp.o
  >>> referenced by ../vm/wks/CMakeFiles/cee_wks_core.dir/__/riscv64/asmhelpers.S.o:(.eh_frame+0x18975A)
  clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
I tried adding its prototype in `exceptionhandling.h` but it was still failing without `DLLEXPORT`.

> This symbol is not meant to be used outside libcoreclr.so.

This visibility is only for pre-linked objects and does not effect what the final libcoreclr.so exports. The libcoreclr.so exports is controlled by `unixexports.src`. i.e. `nm -D libcoreclr.so` does not show this symbol among `coreclr_initialize`, `MetaDataGetDispenser` etc.

Reply:
I think explicit `__attribute__ ((visibility("default")))` would be better since it has nothing to do with exporting the entrypoint from the .so.

Reply:
OK. I saw that `DLLEXPORT` is defined as: `#define DLLEXPORT __attribute__((visibility("default")))`. I'll inline it.

```suggestion
__attribute__((visibility("default"))) EXTERN_C _Unwind_Reason_Code
```